crates/trigger-http/src/server.rs

```diff
@@ -138,12 +148,43 @@ impl<F: RuntimeFactors> HttpServer<F> {
 
     /// Serve incoming requests over the provided [`TcpListener`].
     pub async fn serve(self: Arc<Self>) -> anyhow::Result<()> {
-        let listener = TcpListener::bind(self.listen_addr).await.with_context(|| {
-            format!(
-                "Unable to listen on {listen_addr}",
-                listen_addr = self.listen_addr
-            )
-        })?;
+        let listener: TcpListener = match TcpListener::bind(self.listen_addr).await {
+            Ok(listener) => listener,
+            Err(err) => {
+                if self.find_free_port && err.kind() == ErrorKind::AddrInUse {
+                    let mut found_listener = None;
+                    for _ in 1..=9 {
+                        let mut addr = self.listen_addr;
+                        addr.set_port(addr.port() + 1);
+
+                        match TcpListener::bind(addr).await {
+                            Ok(listener) => {
+                                found_listener = Some(listener);
+                                break;
+                            }
+                            Err(err) => {
+                                if err.kind() == ErrorKind::AddrInUse {
+                                    continue;
+                                }
+                                return Err(anyhow::anyhow!("Unable to listen on {}", addr));
```

Comment:
If there's a non-AddrInUse error on one of the scanned ports, this will produce a strange error message e.g. the user said to listen on port 4000 and gets an error about port 4007.